proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -371,18 +374,393 @@ lemma decode_sc_ctrl_inv_corres:
+           apply (clarsimp simp: ntfn_relation_def split: Structures_A.ntfn.splits)
+          apply (fold updateSchedContext_def)
+          apply (rule updateSchedContext_corres)
+            apply (clarsimp simp: opt_map_red obj_at_simps is_sc_obj)
+            apply (drule (1) pspace_relation_absD[OF _ state_relation_pspace_relation])
+            apply (clarsimp simp: sc_relation_def)
+           apply (clarsimp simp: opt_map_red obj_at_simps is_sc_obj)
+           apply (drule (1) sc_replies_relation_prevs_list'[OF state_relation_sc_replies_relation])
+           apply (clarsimp simp: opt_map_red)
+          apply (clarsimp simp: objBits_simps')
+         apply wpsimp
+        apply wpsimp
+       apply simp+
+     apply wpsimp+
+   apply (fastforce simp: obj_at_simps sc_ntfn_sc_at_def is_ntfn is_sc_obj valid_obj_def)
+  apply clarsimp
+  done
+
+crunches tcb_sched_action, complete_yield_to, reschedule_required, sched_context_resume
+  for in_user_frame[wp]: "in_user_frame buf"
+  (simp: crunch_simps wp: crunch_wps ignore: set_tcb_obj_ref)
+
+lemma
+  schedContext_valid_ipc_buffer_ptr'[wp]:
+  "setSchedContext scp sc \<lbrace>valid_ipc_buffer_ptr' x\<rbrace>" and
+  threadSet_valid_ipc_buffer_ptr'[wp]:
+  "threadSet f tp \<lbrace>valid_ipc_buffer_ptr' x\<rbrace>" and
+  modifyReadyQueuesL1Bitmap_valid_ipc_buffer_ptr'[wp]:
+  "modifyReadyQueuesL1Bitmap d g \<lbrace>valid_ipc_buffer_ptr' x\<rbrace>" and
+  modifyReadyQueuesL2Bitmap_valid_ipc_buffer_ptr'[wp]:
+  "modifyReadyQueuesL2Bitmap d n g \<lbrace>valid_ipc_buffer_ptr' x\<rbrace>" and
+  setQueue_valid_ipc_buffer_ptr'[wp]:
+  "setQueue d prio p \<lbrace>valid_ipc_buffer_ptr' x\<rbrace>"
+  unfolding valid_ipc_buffer_ptr'_def
+  by wpsimp+
+
+crunches schedContextCompleteYieldTo, tcbSchedEnqueue, tcbSchedDequeue,
+         rescheduleRequired, schedContextResume
+  for valid_ipc_buffer_ptr'[wp]: "valid_ipc_buffer_ptr' buf"
+  and sc_at'[wp]: "sc_at' scp"
+  and cur_tcb'[wp]: cur_tcb'
+  (simp: crunch_simps wp: crunch_wps threadSet_cur ignore: setSchedContext)
+
+lemma threadSet_tcbYieldTo_update_valid_queues:
+  "threadSet (tcbYieldTo_update f) t \<lbrace>valid_queues\<rbrace>"
+  apply (simp add: threadSet_def)
+  apply wp
+   apply (simp add: valid_queues_def valid_queues_no_bitmap_def' pred_tcb_at'_def)
+   apply (wp hoare_Ball_helper
+             hoare_vcg_all_lift
+             setObject_tcb_strongest)[1]
+  apply (wp getObject_tcb_wp)
+  apply (clarsimp simp: valid_queues_def valid_queues_no_bitmap_def' pred_tcb_at'_def)
+  apply (fastforce simp: obj_at'_def projectKOs inQ_def)
+  done
+
+lemma sc_yield_from_update_sc_tcb_sc_at[wp]:
+  "set_sc_obj_ref sc_yield_from_update scp ptr \<lbrace>sc_tcb_sc_at P scp\<rbrace>"
+  apply (wpsimp wp: update_sched_context_wp)
+  by (clarsimp simp: sc_tcb_sc_at_def obj_at_def)
+
+lemma schedContextYieldTo_corres:
+  "corres dc ((case_option \<top> in_user_frame buf and einvs) and (\<lambda>s. bound_yt_tcb_at ((=) None) (cur_thread s) s \<and>
```

Comment:
Maybe we should decide on a way to indent the guards and functions in `corres` rules, or at least narrow it down to a couple options

Reply:
Yeah, agreed. It would be nice to not have to think about every time it comes up.

Reply:
In this specific case, the indentation within the preconditions seems a bit off as well and I think that some of the bracketing can be simplified.

Reply:
The brackets are necessary for the correct pattern matching.